pkg/daemon/update.go

```diff
@@ -512,25 +539,35 @@ func (dn *Daemon) logSystem(format string, a ...interface{}) {
 	}
 }
 
-// reboot is the final step. it tells systemd-logind to reboot the machine,
-// cleans up the agent's connections, and then sleeps for 7 days. if it wakes up
-// and manages to return, it returns a scary error message.
-func (dn *Daemon) reboot(rationale string) error {
+const rebootLockPath = "/run/machine-config-daemon-reboot"
+
+// The main idea here is that we *queue* a reboot in a separate systemd unit,
+// which will start once this process exits.  In the orchestrated (not-onceFrom)
+// case, the drain should kill our pod, releasing our lock.  In the onceFrom
+// case we just exit.
+func (dn *Daemon) queueReboot(rationale string) error {
 	// We'll only have a recorder if we're cluster driven
 	if dn.recorder != nil {
 		dn.recorder.Eventf(&corev1.Node{ObjectMeta: metav1.ObjectMeta{Name: dn.name}}, corev1.EventTypeNormal, "Reboot", rationale)
 	}
-	dn.logSystem("machine-config-daemon initiating reboot: %s", rationale)
-
-	// reboot
-	dn.loginClient.Reboot(false)
+	dn.logSystem("machine-config-daemon: initiating reboot: %s", rationale)
 
-	// cleanup
-	dn.Close()
+	fh, err := os.OpenFile(rebootLockPath, os.O_CREATE|os.O_RDONLY, os.FileMode(0600))
+	if err != nil {
+		return err
+	}
 
-	// cross fingers
-	time.Sleep(24 * 7 * time.Hour)
+	if err = syscall.Flock(int(fh.Fd()), syscall.LOCK_EX); err != nil {
+		return errors.Wrapf(err, "flock(%s)", rebootLockPath)
+	}
 
-	// if everything went well, this should be unreachable.
-	return fmt.Errorf("Reboot failed; this error should be unreachable, something is seriously wrong")
+	// We invoke via systemd-run so that this pod can be safely drained
+	rebootCmd := exec.Command("systemd-run", "--unit", "machine-config-daemon-reboot",
+		"--description", fmt.Sprintf("machine-config-daemon: %s", rationale),
+		"flock", rebootLockPath, "reboot")
```

Comment:
A reboot inhibitor could also work here and you wouldn't have to manage that lock file. It would also be a bit more discoverable to someone debugging. This approach seems fine though.